eras/conway/impl/test/Test/Cardano/Ledger/Conway/GovSpec.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Conway.GovSpec where
+
+import Cardano.Ledger.BaseTypes
+import Cardano.Ledger.Conway
+import Cardano.Ledger.Conway.Governance
+import Cardano.Ledger.Shelley.LedgerState
+import Data.Default.Class (Default (..))
+import Data.Maybe
+import Lens.Micro
+import Test.Cardano.Ledger.Common
+import Test.Cardano.Ledger.Conway.ImpTest
+
+spec :: Spec
+spec =
+  describe "Proposals always have valid previous actions" $ do
+    context "Invalid proposals are rejected" $ do
+      itM @Conway "Lookup snapshots" $ do
+        constitutionHash <- freshSafeHash
+        let constitutionAction =
+              NewConstitution
+                SNothing
+                ( Constitution
+                    ( Anchor
+                        (fromJust $ textToUrl "constitution.0")
+                        constitutionHash
+                    )
+                    SNothing
+                )
+        gaidConstitutionProp <- submitProposal constitutionAction
+        let constitutionActionNext =
+              NewConstitution
+                (SJust $ PrevGovActionId (gaidConstitutionProp {gaidGovActionIx = GovActionIx 1})) -- make it wrong
+                ( Constitution
+                    ( Anchor
+                        (fromJust $ textToUrl "constitution.1")
+                        constitutionHash
+                    )
+                    SNothing
+                )
+        submitFailingProposal constitutionActionNext
+      itM @Conway "Lookup snapshots" $ do
+        constitutionHash <- freshSafeHash
+        let constitutionAction =
+              NewConstitution
+                SNothing
+                ( Constitution
+                    ( Anchor
+                        (fromJust $ textToUrl "constitution.0")
+                        constitutionHash
+                    )
+                    SNothing
+                )
+        gaidConstitutionProp <- submitProposal constitutionAction
+        modifyNES $ \nes ->
+          nes
+            & nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . cgEnactStateL . ensPrevConstitutionL
+              .~ SJust (PrevGovActionId gaidConstitutionProp) -- Add the wrong one to PrevGovActionIds
+            & nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . cgGovSnapshotsL
+              .~ def -- Remove from snapshots
```

Comment:
You should not need to modify the new epoch state for these tests
 ```suggestion
```

Reply:
The thought process behind this was the following:
1. In the first test, the initial proposal has been only added to the snapshots and hasn't been enacted yet. We propose an invalid proposal, and expect it to fail.
2. In the second test, the initial proposal has also been only added to the snapshots and hasn't been enacted yet. We propose a valid proposal this time, but we want to make sure that if a previous proposal does not appear in the snapshots but has been enacted before, we still want it to pass the validation check, unless I'm completely misunderstanding this.

I also think I named the tests badly in this case. I will make this more explicit if you are okay with this thought process for testing.

1. Reject when id is not present in either snapshot or enacted state
2. Reject when id is present but purpose does not match in either snapshot or enacted state
3. Accept when id and purpose match in snapshots
4. Accept when id and purpose match in enacted state.

Reply:
I have also now added the test for "first ever proposal" while all other proposals with same purpose have to fail.